src/index.ts

```diff
@@ -441,6 +442,71 @@ export default class FireStash extends EventEmitter {
     this.eventsTimer = null;
   }
 
+  async onThrottledSnapshot(
+    documentPath: string, // app/config
+    callback: (snapshot: FirebaseFirestore.DocumentSnapshot) => void,
+    timeout = 1000,
+  ): Promise<() => void> {
+    return new Promise((resolve, reject) => {
+      let timeoutWindowStart = Date.now();
+      let callsThisTimeout = 0;
+      let numNoChangeTimeouts = 0;
+      let onSnapshotListener: (() => void) | null = null;
+      let interval: NodeJS.Timeout | null = null;
+      let lastUpdateTime = 0;
+
+      const handleSnapshot = async(snapshot: FirebaseFirestore.DocumentSnapshot) => {
+        const now = Date.now();
+        const updateTime = snapshot.updateTime?.toMillis() || 0;
+        const changed = updateTime !== lastUpdateTime;
+        lastUpdateTime = updateTime;
+
+        if (onSnapshotListener) {
+          if (now - timeoutWindowStart >= timeout) {
+            timeoutWindowStart = now;
+            callsThisTimeout = 0;
+          }
+          else callsThisTimeout += 1;
+          if (callsThisTimeout >= 3) {
+            // Cancel the snapshot listener.
+            onSnapshotListener();
+            onSnapshotListener = null;
+            callsThisTimeout = 0;
+            // Switch to polling.
+            interval = setInterval(() => this.db.doc(documentPath).get().then(handleSnapshot, reject), timeout);
+          }
+        }
+        else {
+          if (!changed) ++numNoChangeTimeouts;
+          if (numNoChangeTimeouts >= 3) {
+            // Cancel the interval.
+            interval && clearInterval(interval);
+            numNoChangeTimeouts = 0;
+            onSnapshotListener = this.db.doc(documentPath).onSnapshot(handleSnapshot, reject);
+          }
+        }
+
+        const callbacks = this.throttledSnapshotCallbacks[documentPath] as Set<(snapshot: FirebaseFirestore.DocumentSnapshot) => void>;
+        callbacks.forEach(callback => {
+          callback(snapshot);
+        });
+      };
+
+      // Set the snapshot listener.
+      if (!(documentPath in this.throttledSnapshotCallbacks)) {
+        this.throttledSnapshotCallbacks[documentPath] = new Set();
+        onSnapshotListener = this.db.doc(documentPath).onSnapshot(handleSnapshot, reject);
+      }
+      this.throttledSnapshotCallbacks[documentPath].add(callback);
+      resolve(() => {
+        this.throttledSnapshotCallbacks[documentPath]?.delete(callback);
+        if (this.throttledSnapshotCallbacks[documentPath]?.length === 0) {
+          onSnapshotListener && onSnapshotListener();
```

Comment:
A few things here: 

1. Delete the entry from throttledSnapshotCallbacks if the set is empty `this.throttledSnapshotCallbacks.delete(documentPath)`
2. We need to store the snapshot listener on the FireStash instance too. This onShapshotListener is specific to the closure, so only the initializing watcher can actually close the connection.
3. We need to also cancel the timeout if it exists here (which also needs to be shared with all invocations on the instance.